File: frontend/ts-mfe-lms/src/types/discussion.ts
export interface DiscussionDTO {
  id: string;
  title: string;
  content: string;
  courseId: string;
  courseTitle: string;
  authorId: string;
  authorName: string;
  authorAvatar: string;
  authorRole: string;
  type: DiscussionType;
  isPinned: boolean;
  isLocked: boolean;
  isResolved: boolean;
  viewCount: number;
  likeCount: number;
  replyCount: number;
  createdAt: string;
  updatedAt: string;
  lastReplyAt?: string;
  lastReplyAuthorName?: string;
  isLikedByCurrentUser: boolean;
  formattedCreatedAt: string;
  formattedUpdatedAt: string;
}

export interface DiscussionReplyDTO {
  id: string;
  content: string;
  discussionId: string;
  authorId: string;
  authorName: string;
  authorAvatar: string;
  authorRole: string;
  parentReplyId?: string;
  parentReplyAuthorName?: string;
  childReplies: DiscussionReplyDTO[];
  likeCount: number;
  isEdited: boolean;
  isInstructorReply: boolean;
  isAcceptedAnswer: boolean;
  createdAt: string;
  updatedAt: string;
  editedAt?: string;
  isLikedByCurrentUser: boolean;
  formattedCreatedAt: string;
  formattedEditedAt: string;
}

export interface CreateDiscussionRequest {
  title: string;
  content: string;
  courseId: string;
  type: DiscussionType;
}

export interface UpdateDiscussionRequest {
  title: string;
  content: string;
}

export interface CreateReplyRequest {
  content: string;
  parentReplyId?: string;
}

export interface UpdateReplyRequest {
  content: string;
}

export interface DiscussionSearchRequest {
  query?: string;
  courseId?: string;
  type?: DiscussionType;
  isResolved?: boolean;
  isPinned?: boolean;
  sortBy: DiscussionSortOrder;
  page: number;
  pageSize: number;
}

export interface DiscussionListResponse {
  discussions: DiscussionDTO[];
  pagination: PaginationMetadata;
  stats: DiscussionStats;
}

export interface DiscussionDetailResponse {
  discussion: DiscussionDTO;
  replies: DiscussionReplyDTO[];
  stats: DiscussionStats;
}

export interface DiscussionStats {
  totalDiscussions: number;
  totalReplies: number;
  resolvedQuestions: number;
  pendingQuestions: number;
  userParticipationCount: number;
  typeDistribution: Record<DiscussionType, number>;
}

export interface PaginationMetadata {
  page: number;
  pageSize: number;
  totalCount: number;
  totalPages: number;
  hasNext: boolean;
  hasPrevious: boolean;
}

export enum DiscussionType {
  General = 0,
  Question = 1,
  Announcement = 2,
  Resource = 3,
  Assignment = 4
}

export enum DiscussionSortOrder {
  Latest = 0,
  Oldest = 1,
  MostReplies = 2,
  MostLikes = 3,
  MostViews = 4,
  PinnedFirst = 5
}

export interface DiscussionActivityDTO {
  id: string;
  type: string;
  title: string;
  content: string;
  courseId: string;
  courseTitle: string;
  authorId: string;
  authorName: string;
  authorAvatar: string;
  createdAt: string;
  formattedCreatedAt: string;
  actionUrl: string;
}

export interface NotificationPreferencesDTO {
  newDiscussionNotifications: boolean;
  replyNotifications: boolean;
  likeNotifications: boolean;
  instructorResponseNotifications: boolean;
  dailyDigest: boolean;
  watchedCourses: string[];
}

export interface ModerationActionDTO {
  discussionId: string;
  replyId?: string;
  action: string;
  reason?: string;
}

File: backends/backend-dotnet/Models/Discussion.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace TalentSphere.API.Models
{
    public class Discussion
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid Id { get; set; }

        [Required]
        [StringLength(200)]
        public string Title { get; set; } = string.Empty;

        [Required]
        public string Content { get; set; } = string.Empty;

        [Required]
        public Guid CourseId { get; set; }
        
        [ForeignKey("CourseId")]
        public Course? Course { get; set; }

        [Required]
        public Guid AuthorId { get; set; }
        
        [ForeignKey("AuthorId")]
        public User? Author { get; set; }

        [Required]
        public DiscussionType Type { get; set; } = DiscussionType.General;

        public bool IsPinned { get; set; } = false;
        
        public bool IsLocked { get; set; } = false;
        
        public bool IsResolved { get; set; } = false;

        public int ViewCount { get; set; } = 0;
        
        public int LikeCount { get; set; } = 0;
        
        public int ReplyCount { get; set; = 0;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime? LastReplyAt { get; set; }
        
        public Guid? LastReplyById { get; set; }
    }

    public class DiscussionReply
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid Id { get; set; }

        [Required]
        public string Content { get; set; } = string.Empty;

        [Required]
        public Guid DiscussionId { get; set; }
        
        [ForeignKey("DiscussionId")]
        public Discussion? Discussion { get; set; }

        [Required]
        public Guid AuthorId { get; set; }
        
        [ForeignKey("AuthorId")]
        public User? Author { get; set; }

        public Guid? ParentReplyId { get; set; }
        
        [ForeignKey("ParentReplyId")]
        public DiscussionReply? ParentReply { get; set; }
        
        public ICollection<DiscussionReply>? ChildReplies { get; set; }

        public int LikeCount { get; set; } = 0;

        public bool IsEdited { get; set; } = false;
        
        public bool IsInstructorReply { get; set; } = false;
        
        public bool IsAcceptedAnswer { get; set; } = false;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
        
        public DateTime? EditedAt { get; set; }
    }

    public class DiscussionLike
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid Id { get; set; }

        [Required]
        public Guid UserId { get; set; }
        
        [ForeignKey("UserId")]
        public User? User { get; set; }

        [Required]
        public Guid DiscussionId { get; set; }
        
        [ForeignKey("DiscussionId")]
        public Discussion? Discussion { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }

    public class ReplyLike
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public Guid Id { get; set; }

        [Required]
        public Guid UserId { get; set; }
        
        [ForeignKey("UserId")]
        public User? User { get; set; }

        [Required]
        public Guid ReplyId { get; set; }
        
        [ForeignKey("ReplyId")]
        public DiscussionReply? Reply { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }

    public enum DiscussionType
    {
        General = 0,
        Question = 1,
        Announcement = 2,
        Resource = 3,
        Assignment = 4
    }
}